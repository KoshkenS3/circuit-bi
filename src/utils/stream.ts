import { isArray } from "lodash";
import { Transform, TransformCallback } from "stream";
import { StringDecoder } from "string_decoder";

// adapted from https://futurestud.io/tutorials/node-js-filter-data-in-streams
export class StreamFilterTransform<RowType> extends Transform {
  constructor(protected includeRow: (row: RowType) => boolean) {
    super({
      readableObjectMode: true,
      writableObjectMode: true,
    });
  }

  _transform(row: RowType, encoding: string, next: TransformCallback): void {
    if (this.includeRow(row)) {
      return next(null, row);
    }
    next();
  }
}

// adapted from https://futurestud.io/tutorials/node-js-filter-data-in-streams
export class FlattenStream<RowType> extends Transform {
  constructor() {
    super({
      readableObjectMode: true,
      writableObjectMode: true,
    });
  }

  _transform(rows: RowType[], encoding: string, next: TransformCallback): void {
    if (!isArray(rows)) {
      throw new Error(`Supplied input rows is not an array`);
    }
    for (const row of rows) {
      this.push(row);
    }
    next();
  }
}

export class StreamBatchBy<RowType> extends Transform {
  protected currentBatchKey: string | number | null = null;
  protected currentBatch: RowType[] = [];

  constructor(protected getBatchKey: (row: RowType) => string | number) {
    super({
      readableObjectMode: true,
      writableObjectMode: true,
    });
  }

  _transform(row: RowType, encoding: string, next: TransformCallback): void {
    const rowKey = this.getBatchKey(row);
    if (this.currentBatchKey === null) {
      this.currentBatchKey = rowKey;
    }
    if (rowKey === this.currentBatchKey) {
      this.currentBatch.push(row);
      return next();
    }
    // flush current batch
    this.push(this.currentBatch);
    this.currentBatch = [row];
    this.currentBatchKey = rowKey;
    next();
  }
}

export class StreamAggBy<RowType> extends Transform {
  protected currentBatchKey: string | number | null = null;
  protected currentBatch: RowType[] = [];

  constructor(
    protected getBatchKey: (row: RowType) => string | number,
    protected aggregate: (rows: RowType[]) => RowType
  ) {
    super({
      readableObjectMode: true,
      writableObjectMode: true,
    });
  }

  _transform(row: RowType, encoding: string, next: TransformCallback): void {
    const rowKey = this.getBatchKey(row);
    if (this.currentBatchKey === null) {
      this.currentBatchKey = rowKey;
    }
    if (rowKey === this.currentBatchKey) {
      this.currentBatch.push(row);
      return next();
    }
    // flush current batch
    const aggregatedRow = this.aggregate(this.currentBatch);
    this.push(aggregatedRow);
    this.currentBatch = [row];
    this.currentBatchKey = rowKey;
    next();
  }
}

export class StreamBufferToLinesTransform extends Transform {
  protected rest = "";
  protected decoder: StringDecoder;

  constructor(protected lineSeparator = "\n", protected encoding = "utf8") {
    super({
      readableObjectMode: true,
      writableObjectMode: true,
    });
    this.decoder = new StringDecoder(encoding);
  }

  _transform(buffer: Buffer, encoding: string, next: TransformCallback): void {
    const lines = (this.rest + this.decoder.write(buffer)).split(this.lineSeparator);
    this.rest = lines.pop() || "";
    for (const line of lines) {
      this.push(line);
    }
    next();
  }
  _flush(next: TransformCallback) {
    const lines = (this.rest + this.decoder.end()).split(this.lineSeparator);
    for (const line of lines) {
      this.push(line);
    }
    this.rest = "";
    next();
  }
}

export class StreamLinesToBufferTransform extends Transform {
  protected lines: string[] = [];

  constructor(protected lineDelimiter: string = "\n", protected lineCount: number = 100) {
    super({
      readableObjectMode: true,
      writableObjectMode: true,
    });
  }

  _transform(line: string, encoding: string, next: TransformCallback): void {
    this.lines.push(line);
    if (this.lines.length >= this.lineCount) {
      this.push(Buffer.from(this.lines.join(this.lineDelimiter) + this.lineDelimiter));
      this.lines = [];
    }
    next();
  }
  _flush(next: TransformCallback) {
    this.push(Buffer.from(this.lines.join(this.lineDelimiter) + this.lineDelimiter));
    this.lines = [];
    next();
  }
}

export class StreamConsoleLogDebug<Input> extends Transform {
  constructor(protected consoleName: string = "line", protected transform: (item: Input) => any = (item) => item) {
    super({
      readableObjectMode: true,
      writableObjectMode: true,
    });
  }

  _transform(item: Input, encoding: string, next: TransformCallback): void {
    console.log({ [this.consoleName]: this.transform(item) });
    next();
  }
}
