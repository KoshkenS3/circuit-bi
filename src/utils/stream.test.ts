import { PassThrough, Readable } from "stream";
import {
  FlattenStream,
  StreamAggBy,
  StreamBatchBy,
  StreamBufferToLinesTransform,
  StreamLinesToBufferTransform,
} from "./stream";

describe("test stream helpers", () => {
  it("FlattenStream: should batch rows by key", () => {
    interface RowType {
      id: string;
      num: number;
    }
    const input = new PassThrough({ objectMode: true });
    const batchBy = new FlattenStream<RowType>();

    const res: RowType[] = [];
    input.pipe(batchBy).on("data", (row) => res.push(row));

    const rows: RowType[][] = Array.from({ length: 3 }).map((_, i) =>
      Array.from({ length: 4 }).map((_, j) => ({
        id: `${i}-${j}`,
        num: i * j,
      }))
    );

    for (const row of rows) {
      input.push(row);
    }
    expect(res.length).toEqual(12);
    expect(res).toMatchSnapshot();
  });

  it("StreamBatchBy: should batch rows by key", () => {
    interface RowType {
      id: string;
      num: number;
    }
    const input = new PassThrough({ objectMode: true });
    const batchBy = new StreamBatchBy<RowType>((row) => row.id);

    const res: RowType[][] = [];
    input.pipe(batchBy).on("data", (rows) => res.push(rows));

    const rows = Array.from({ length: 10 }).map((_, i) => ({
      id: `id${Math.floor(i / 3)}`,
      num: i,
    }));

    for (const row of rows) {
      input.push(row);
    }
    expect(res.length).toEqual(3); // we don't yield the last batch as we don't know if it's finished
    expect(res).toMatchSnapshot();
  });

  it("StreamAggBy: should aggregate rows by key", () => {
    interface RowType {
      id: string;
      num: number;
    }
    const input = new PassThrough({ objectMode: true });
    const batchBy = new StreamAggBy<RowType>(
      (row) => row.id,
      (rows) => ({
        id: rows[0].id,
        num: rows.reduce((acc, row) => acc + row.num, 0),
      })
    );

    const res: RowType[][] = [];
    input.pipe(batchBy).on("data", (rows) => res.push(rows));

    const rows = Array.from({ length: 10 }).map((_, i) => ({
      id: `id${Math.floor(i / 3)}`,
      num: i,
    }));

    for (const row of rows) {
      input.push(row);
    }
    expect(res.length).toEqual(3); // we don't yield the last batch as we don't know if it's finished
    expect(res).toMatchSnapshot();
  });

  it("StreamBufferToLinesTransform: parses buffer to lines properly", () => {
    const buffer = Buffer.from("hello\nworld\nlorem ipsum\nbut not this one");
    const toLines = new StreamBufferToLinesTransform();

    const res: string[] = [];
    toLines.on("data", (line) => res.push(line));

    toLines.write(buffer);

    // we don't want to yield last one yet as it may be truncated
    expect(res.length).toBe(3);
    expect(res).toMatchSnapshot();

    toLines.end();

    // now we expect the last one
    expect(res.length).toBe(4);
    expect(res).toMatchSnapshot();
  });

  it("StreamLinesToBufferTransform: should transform lines to buffer", () => {
    const lines = "hello\nworld\nlorem ipsum\nlorem ipsum\none more line";
    const toBuffer = new StreamLinesToBufferTransform("\n", 2 /* line count */);

    const res: Buffer[] = [];
    toBuffer.on("data", (buffer) => res.push(buffer));

    for (const line of lines.split("\n")) {
      toBuffer.write(line);
    }

    expect(res.length).toBe(2);
    expect(Buffer.concat(res).toString("utf8")).toMatchSnapshot();

    toBuffer.end();
    // now we expect the last one
    expect(res.length).toBe(3);
    expect(Buffer.concat(res).toString("utf8")).toMatchSnapshot();
  });
});
