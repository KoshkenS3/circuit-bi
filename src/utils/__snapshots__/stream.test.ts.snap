// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test stream helpers FlattenStream: should batch rows by key 1`] = `
Array [
  Object {
    "id": "0-0",
    "num": 0,
  },
  Object {
    "id": "0-1",
    "num": 0,
  },
  Object {
    "id": "0-2",
    "num": 0,
  },
  Object {
    "id": "0-3",
    "num": 0,
  },
  Object {
    "id": "1-0",
    "num": 0,
  },
  Object {
    "id": "1-1",
    "num": 1,
  },
  Object {
    "id": "1-2",
    "num": 2,
  },
  Object {
    "id": "1-3",
    "num": 3,
  },
  Object {
    "id": "2-0",
    "num": 0,
  },
  Object {
    "id": "2-1",
    "num": 2,
  },
  Object {
    "id": "2-2",
    "num": 4,
  },
  Object {
    "id": "2-3",
    "num": 6,
  },
]
`;

exports[`test stream helpers StreamAggBy: should aggregate rows by key 1`] = `
Array [
  Object {
    "id": "id0",
    "num": 3,
  },
  Object {
    "id": "id1",
    "num": 12,
  },
  Object {
    "id": "id2",
    "num": 21,
  },
]
`;

exports[`test stream helpers StreamBatchBy: should batch rows by key 1`] = `
Array [
  Array [
    Object {
      "id": "id0",
      "num": 0,
    },
    Object {
      "id": "id0",
      "num": 1,
    },
    Object {
      "id": "id0",
      "num": 2,
    },
  ],
  Array [
    Object {
      "id": "id1",
      "num": 3,
    },
    Object {
      "id": "id1",
      "num": 4,
    },
    Object {
      "id": "id1",
      "num": 5,
    },
  ],
  Array [
    Object {
      "id": "id2",
      "num": 6,
    },
    Object {
      "id": "id2",
      "num": 7,
    },
    Object {
      "id": "id2",
      "num": 8,
    },
  ],
]
`;

exports[`test stream helpers StreamBufferToLinesTransform: parses buffer to lines properly 1`] = `
Array [
  "hello",
  "world",
  "lorem ipsum",
]
`;

exports[`test stream helpers StreamBufferToLinesTransform: parses buffer to lines properly 2`] = `
Array [
  "hello",
  "world",
  "lorem ipsum",
  "but not this one",
]
`;

exports[`test stream helpers StreamLinesToBufferTransform: should transform lines to buffer 1`] = `
"hello
world
lorem ipsum
lorem ipsum
"
`;

exports[`test stream helpers StreamLinesToBufferTransform: should transform lines to buffer 2`] = `
"hello
world
lorem ipsum
lorem ipsum
one more line
"
`;
